### DISCUSSION TITLE
Is there a replacement for distutils.command.clean with setuptools >= 49.1.2?

### DISCUSSION THREAD

**Discussion by DougBurke** [pypa/setuptools#2838] [Q&A] [answered]
I am following https://setuptools.pypa.io/en/latest/deprecated/distutils-legacy.html trying to update a package which includes

```
from distutils.command.clean import clean as _clean
```

The best I've come up with is the following, but is there a better way to do this?

```
# Is there a better way to get hold of this?
from setuptools._distutils.command.clean import clean as _clean
```

The python package in question builds some compiled code using configure and so it hooks into `clean` to remove the debris from that step. I would prefer removing this override, since `git clean -dfx` is likely a more-robust solution, but someone may be using it so if I can say "using clean is a bad idea nowadays" I have some justification!

**Comment by webknjaz** [comment, marked as answer]
Invoking `setup.py` directly is highly discouraged: https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html. Would this explanation be enough?

**Reply by webknjaz** [reply]
Also, you should probably move this discussion to the Q&A category.

**Reply by DougBurke** [reply]
Ta for the link. So, the discussion at  https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.htm includes the text

```
Right now the most common targets have replacements: install, sdist, develop [5] and bdist_wheel all have standards backing them, and test can be replaced by tox, nox or even a Makefile. If you use one of the more obscure commands and cannot find a replacement, it would be good to raise that on the packaging discourse to see what replacements are or could be made available.
```

I claim that this is my attempt at trying to find out what the replacement is. I'm using this venue rather than the discourse because https://setuptools.pypa.io/en/latest/deprecated/distutils-legacy.html told me to come here.

So it would be helpful to know if there is going to be a replacement for `clean` in some manner or whether it is going to be removed (reading between the lines of discussions on editable install I assume the clean option is not likely to be kept around).

[We also have commands that we've added to configure the extensions we are building, but I'm trying to understand the easy ones first]

**Reply by webknjaz** [reply]
> So it would be helpful to know if there is going to be a replacement for `clean` in some manner or whether it is going to be removed (reading between the lines of discussions on editable install I assume the clean option is not likely to be kept around).
>
> [We also have commands that we've added to configure the extensions we are building, but I'm trying to understand the easy ones first]

I don't think it makes sense given that PEP 517 doesn't have any protocol for this. Moreover, the build env setup is usually up to the build front-end (`pip` or `pypa/build`, for example) and they often just create "throw-away" virtualenvs under `/tmp` that you wouldn't need to clean up. I know that `pip` is experimenting with having in-tree builds, though but dunno all that implications. Either way, I'd expect this to need an interface on the front-end side rather than a backend (which `setuptools` is).
PEP 517 has a mechanism for passing free-form args additionally but the hooks for building wheels and sdists would normally get those and there's no "cleanup" hook in the standard. Besides, `setuptools` doesn't currently implement this mechanism anyway, and `pip` doesn't as well. AFAIK only `build` has support for the custom args.
Note, that in the modern world, you should use `build` for your packaging process. Some people used to abuse `setup.py` to be an all-in-one arbitrary commands runner which should've never been possible in the first place. If you want that, there are tools like tox/nox/make that are better suited anyway.
As for the customizations around building extensions, that will probably still work (depending on your setup). The problem is that setuptools grew w/o any standards and different people probably ended up putting all sorts of code there turning it into a generic task runner. I think it should still be possible to integrate it all so that it's called by the hooks since the distutils/setuptools commands can call other distutils/setuptools commands.
But if there's something very different from the rest of the ecosystem, you may end up writing your own PEP517 in-tree backend just for your project.

**Reply by DougBurke** [reply]
For the clean case I have decided to just remove that command. The remaining commands I'm leaving for now as they extent numpy.distutils commands and I honestly don't have the time to understand what the current state is there (and I definitely don't want to have to go to the effort of learning how to build a back end myself ;-)

Thanks for the input.
